File: SourceView/ChildEditViewController.m
/*
 Copyright (C) 2016 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 View controller object for the edit bookmark sheet.
 */

#import "ChildEditViewController.h"

@interface ChildEditViewController ()

@property (nonatomic, weak) IBOutlet NSButton *doneButton;
@property (nonatomic, weak) IBOutlet NSTextField *nameField;
@property (nonatomic, weak) IBOutlet NSTextField *urlField;

@end


#pragma mark -

@implementation ChildEditViewController

// -------------------------------------------------------------------------------
//	viewWillAppear
// -------------------------------------------------------------------------------
- (void)viewWillAppear
{
    [super viewWillAppear];
 
    self.nameField.stringValue = self.savedValues[kName_Key];
    self.urlField.stringValue = self.savedValues[kURL_Key];
    self.doneButton.enabled = [self doneAllowed];
}

// -------------------------------------------------------------------------------
//	doneAllowed
// -------------------------------------------------------------------------------
- (BOOL)doneAllowed
{
    return (self.nameField.stringValue.length > 0 && self.urlField.stringValue.length > 0);
}

// -------------------------------------------------------------------------------
//	done:sender
// -------------------------------------------------------------------------------
- (IBAction)done:(id)sender
{
    // add the http prefix if the user forgot to
    NSString *urlStr;
    if (![self.urlField.stringValue hasPrefix:HTTP_PREFIX])
    {
        urlStr = [NSString stringWithFormat:@"%@%@", HTTP_PREFIX, self.urlField.stringValue];
    }
    else
    {
        urlStr = self.urlField.stringValue;
    }
    _savedValues = [NSMutableDictionary dictionaryWithObjectsAndKeys:
                        self.nameField.stringValue, kName_Key,
                        [NSURL URLWithString:urlStr], kURL_Key,
                   nil];
    [self clearValues];
    
    [self.view.window.sheetParent endSheet:self.view.window returnCode:NSModalResponseOK];
}

// -------------------------------------------------------------------------------
//	clearValues
// -------------------------------------------------------------------------------
- (void)clearValues
{
    self.nameField.stringValue = self.urlField.stringValue = @"";
}

// -------------------------------------------------------------------------------
//	cancel:sender
// -------------------------------------------------------------------------------
- (IBAction)cancel:(id)sender
{
    [self clearValues];
    [self.view.window.sheetParent endSheet:self.view.window returnCode:NSModalResponseCancel];
}

// -------------------------------------------------------------------------------
//	controlTextDidChange:obj
//
//  For this to be called, we need to be a delegate to both NSTextFields
// -------------------------------------------------------------------------------
- (void)controlTextDidChange:(NSNotification *)obj
{
    self.doneButton.enabled = [self doneAllowed];
}

@end

File: SourceView/ChildEditViewController.swift
//
//  ChildEditViewController.swift
//  SourceView
//
//  Translated by OOPer in cooperation with shlab.jp, on 2015/10/24.
//
//
/*
 Copyright (C) 2018 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information

 Abstract:
 View controller object for the edit bookmark sheet.
 */
import Cocoa

// keys to use to obtain the name and url values from the returned NSDictionary
let kName_Key = "name"
let kURL_Key = "url"

@objc(ChildEditViewController)
class ChildEditViewController: NSViewController {
    
    var savedValues: [String: Any] = [:]
    
    @IBOutlet private weak var doneButton: NSButton!
    @IBOutlet private weak var nameField: NSTextField!
    @IBOutlet private weak var urlField: NSTextField!
    
    
    //MARK: -
    
    // -------------------------------------------------------------------------------
    //	viewWillAppear
    // -------------------------------------------------------------------------------
    override func viewWillAppear() {
        super.viewWillAppear()
        
        self.nameField.stringValue = self.savedValues[kName_Key] as! String? ?? ""
        self.urlField.stringValue = String(describing: self.savedValues[kURL_Key] ?? "")
        self.doneButton.isEnabled = self.doneAllowed
    }
    
    // -------------------------------------------------------------------------------
    //	doneAllowed
    // -------------------------------------------------------------------------------
    private var doneAllowed: Bool {
        return (!self.nameField.stringValue.isEmpty && !self.urlField.stringValue.isEmpty)
    }
    
    // -------------------------------------------------------------------------------
    //	done:sender
    // -------------------------------------------------------------------------------
    @IBAction func done(_: AnyObject) {
        let urlStr: String
        if !self.urlField.stringValue.hasPrefix(HTTP_PREFIX) {
            urlStr = "\(HTTP_PREFIX)\(self.urlField.stringValue)"
        } else {
            urlStr = self.urlField.stringValue
        }
        savedValues = [
            kName_Key : self.nameField.stringValue as AnyObject,
            kURL_Key : URL(string: urlStr)! as AnyObject
        ]
        self.clearValues()
        
        self.view.window?.sheetParent?.endSheet(self.view.window!, returnCode: NSApplication.ModalResponse.OK)
    }
    
    // -------------------------------------------------------------------------------
    //	clearValues
    // -------------------------------------------------------------------------------
    private func clearValues() {
        self.nameField.stringValue = ""
        self.urlField.stringValue = ""
    }
    
    // -------------------------------------------------------------------------------
    //	cancel:sender
    // -------------------------------------------------------------------------------
    @IBAction func cancel(_: AnyObject) {
        self.clearValues()
        self.view.window?.sheetParent?.endSheet(self.view.window!, returnCode: NSApplication.ModalResponse.cancel)
    }
    
    // -------------------------------------------------------------------------------
    //	controlTextDidChange:obj
    //
    //  For this to be called, we need to be a delegate to both NSTextFields
    // -------------------------------------------------------------------------------
    func controlTextDidChange(_ obj: Notification) {
        self.doneButton.isEnabled = self.doneAllowed
    }
    
}
